Update minimum msrestazure package to 0.4.9

action opened the pull request:
The Azure SDK for Python, 2.0.0rc6, has an implicit dependency on
python module requests 2.8.0+.

The msrestazure module 0.4.9 changes its dependency on msrest from
0.4.4 to 0.4.10; however, the Azure SDK for Python defines its
minimum version of msrestazure as 0.4.0.

Further, version 0.4.8 of the msrest module changes its minimum
version requirement of requests from 2.7.0 to 2.14.1.

If one uses the request module prior to 2.8.0 then they are likely
to see this exception:
  SysCallError: (32, 'EPIPE')

The exception can be triggered by doing the following:
  1. Use the Azure SDK to store a copy of the compute and network clients
  2. Create a VM
  3. Create a NIC
  4. Deallocate VM and attach NIC to VM
  5. Start VM
  6. The next use of the compute or network clients will trigger a
     SysCallError: (32, 'EPIPE') error.

The EPIPE error is gracefully handled by the requests module in 2.8.0.

Without this change, any user on a system that is running an older
version of requests and installs the 2.0.0rc6 Azure SDK via pip
will encounter this issue because the minimum dependencies will be
satisfied and requests will not be upgraded.

msftclas commented:
This seems like a small (but important) contribution, so no Contribution License Agreement is required at this point. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

lmazuel commented:
Hi @action 

This file is not used in our packaging system, this is only for Travis.

I'm not sure what is your issue, but we strongly recommend to use a virtualenv and not the system directly. This way you are sure to get the latest requests each time.

action commented:
I see. To resolve this issue during install, to me it looks like I need to move this change to azure-common/setup.py. If it's dependency on msrestazure is updated, then future versions of the SDK will install the correct version of requests if an outdated one already exists in an non-virtualenv deployment.

I did not intend to change the Travis workflow, so go ahead and reject this change. It appears to resolve the problem, a small change is required to a different file.

action commented:
Actually, I updated this pull request to make the change from requirements.txt to azure-common/setup.py. This request shouldn't be rejected, and I've moved the change from the Travis process to the install process. The updated change should cause fresh installs of the package to avoid the situation in which an old, and incompatible, version of requests erroneously satisfies the package dependency.

lmazuel commented:
Actually, in 2.0.0rc6 msrestazure was a dependency of azure-common as "extra_requires":
```python
    extras_require={
        'autorest':[
            'msrestazure>=0.4.0,<0.5.0',
        ]
    },
```
But we got issue with [pip and recursive extra requires](https://github.com/Azure/azure-sdk-for-python/issues/644).

So this **not** used anymore, Now, every package has their own msrestazure dependency. To avoid mistake on duplication, we have a [template of package](https://github.com/Azure/cookiecutter-azuresdk-pypackage/) and we use it each time we do a release.

The template currently contains:
```python
    install_requires=[
        'msrestazure~=0.4.8',
        'azure-common~=1.1.6',
    ],
```
So I can update the template to 0.4.11, this way new people will get a new dependency, but for new package only.

Talking about 2.0.0rc6, the dependency was `msrestazure>=0.4.0,<0.5.0`. So, a fresh install should pick latest msrestazure, then latest msrest, then apply `requests~=2.14`. Unless you have old version of msrest/msrestazure in your pip cache.

action commented:
Interesting. Thanks for the detailed reply, I found it very helpful.

I'm satisfied knowing that the issue has been looked into and should be resolved for future releases of the SDK. And, yea, there might have been a chance an older version of msrest/msrestazure was in my pip cache.
